coreMacros/src/main/scala/Chisel/internal/sourceinfo/SourceInfoTransform.scala

```diff
@@ -0,0 +1,180 @@
+  import c.universe._
+  def apply[T: c.WeakTypeTag](size: c.Tree, t: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($size, $t)($implicitSourceInfo)"
+  }
+}
+
+class RegTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply[T: c.WeakTypeTag](t: c.Tree): c.Tree = {
+    val tpe = weakTypeOf[T]
+    q"$thisObj.do_apply($t, null.asInstanceOf[$tpe], null.asInstanceOf[$tpe])($implicitSourceInfo)"
+  }
+}
+
+class WhenTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply(cond: c.Tree)(block: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($cond, $block)($implicitSourceInfo)"
+  }
+
+  def elsewhen(elseCond: c.Tree)(block: c.Tree): c.Tree = {
+    q"$thisObj.do_elsewhen($elseCond, $block)($implicitSourceInfo)"
+  }
+
+  def otherwise(block: c.Tree): c.Tree = {
+    q"$thisObj.do_otherwise($block)($implicitSourceInfo)"
+  }
+}
+
+class StopTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def stop(): c.Tree = {
+    q"$thisObj.do_stop($implicitSourceInfo)"
+  }
+}
+
+class PrintTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply(fmt: c.Tree, data: c.Tree*): c.Tree = {
+    q"$thisObj.do_apply($fmt, ..$data)($implicitSourceInfo)"
+  }
+}
+
+class AssertTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply_msg(cond: c.Tree, message: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($cond, $message)($implicitSourceInfo)"
+  }
+  def apply(cond: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($cond)($implicitSourceInfo)"
+  }
+}
+
+class MuxTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply[T: c.WeakTypeTag](cond: c.Tree, con: c.Tree, alt: c.Tree): c.Tree = {
+    val tpe = weakTypeOf[T]
+    q"$thisObj.do_apply[$tpe]($cond, $con, $alt)($implicitSourceInfo)"
+  }
+}
+
+class VecTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply_elts(elts: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($elts)($implicitSourceInfo)"
+  }
+  def apply_elt0(elt0: c.Tree, elts: c.Tree*): c.Tree = {
+    q"$thisObj.do_apply($elt0, ..$elts)($implicitSourceInfo)"
+  }
+  def tabulate(n: c.Tree)(gen: c.Tree): c.Tree = {
+    q"$thisObj.do_tabulate($n)($gen)($implicitSourceInfo)"
+  }
+  def fill(n: c.Tree)(gen: c.Tree): c.Tree = {
+    q"$thisObj.do_fill($n)($gen)($implicitSourceInfo)"
+  }
+  def contains(x: c.Tree)(ev: c.Tree): c.Tree = {
+    q"$thisObj.do_contains($x)($implicitSourceInfo, $ev)"
+  }
+}
+
+/** "Automatic" source information transform / insertion macros, which generate the function name
+  * based on the macro invocation (instead of explicitly writing out every transform).
+  */
+class SourceInfoTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  /** Returns the TermName of the transformed function, which is the applied function name with do_
```

Comment:
Did you mean writing explicit transform for all cases, or just for cases where two API functions that invoke the same thing?

I think the first would require way too much boilerplate. Yes, just one transform line per functions, but there are a lot of functions. All the logic operators, arithmetic operators, those operators in non-bit-growth form, and so on...

If the second, while the total lines of code is approximately a wash, I think the style and consistency gains are worth it. Sure, there are extra function calls, but I also think if we ever stack trace in a Chisel core do_, then that's probably not on the user to debug, and it's only one extra line. It's also no worse than the current setup, and it may also provide debugging message readability gains in that it explicitly lists (for example) do_!= is called, which chains into a call to do_=/=, instead of just a call to do_=/= which could lead users to wonder where the do_=/= came from when they only used !=.